### PR TITLE
Support for tcsh completion

### PR DESCRIPTION
**Pull request by simaoafonso-pwt**
This is incomplete due to tcsh limitations, but functional.

See #56

Here are some examples generated with this:

```tcsh
#!/usr/bin/env tcsh
# AUTOMATICALLY GENERATED by `shtab`


# Custom Preamble
set autolist
# End Custom Preamble


complete customcomplete \
        'c/--/(output-name input-file help)/' \
        'c/-/(h i o -)/' \
        'n/-i/f/' \
        'n/--input-file/f/' \
        'n/-o/d/' \
        'n/--output-name/d/' \
        'p/1/(completion process)/' \
        'p@2@`set cmd=($COMMAND_LINE); ( [ "$cmd[2]" == "completion" ] && echo "bash\nzsh\ntcsh" || false ) || ( echo "oa\nob" || false )`@' \
        'p/*/()/'
#!/usr/bin/env tcsh
# AUTOMATICALLY GENERATED by `shtab`



complete pathcomplete \
        'c/--/(help dir print-completion)/' \
        'c/-/(s h -)/' \
        'n/-s/(bash zsh tcsh)/' \
        'n/--print-completion/(bash zsh tcsh)/' \
        'n/--dir/d/' \
        'p/1/f/' \
        'p/*/()/'
#!/usr/bin/env tcsh
# AUTOMATICALLY GENERATED by `shtab`



complete shtab \
        'c/--/(error-unimportable shell preamble verbose prog prefix help version)/' \
        'c/-/(s v u h -)/' \
        'n/-s/(bash zsh tcsh)/' \
        'n/--shell/(bash zsh tcsh)/' \
        'p/*/()/'
```

### CODE REVIEW

**Comment by simaoafonso-pwt** [comment]
Pinging @casperdcl from #56.

**Comment by simaoafonso-pwt** [comment]
Friendly ping @casperdcl

File: shtab/__init__.py [hunk 1]
```diff
@@ -39,8 +40,16 @@
...
-    "directory": {"bash": "_shtab_compgen_dirs", "zsh": "_files -/"},
+    "file": {
+        "bash": "_shtab_compgen_files",
+        "zsh": "_files",
+        "tcsh": "f",  # Not strictly a function...
+    },
+    "directory": {
+        "bash": "_shtab_compgen_dirs",
+        "zsh": "_files -/",
+        "tcsh": "d",  # Not strictly a function...
```
Comment (casperdcl) [inline shtab/__init__.py:51]:
```suggestion
        "tcsh": "f",
    },
    "directory": {
        "bash": "_shtab_compgen_dirs",
        "zsh": "_files -/",
        "tcsh": "d",
```

File: shtab/main.py [hunk 2]
```diff
@@ -34,12 +34,16 @@ def get_main_parser():
         action="store_true",
         help="raise errors if `parser` is not found in $PYTHONPATH",
     )
+    parser.add_argument("-v", "--verbose", dest="loglevel",
```
Comment (casperdcl) [inline shtab/main.py:37]:
```suggestion
    parser.add_argument("--verbose", dest="loglevel",
```

to avoid potential confusion with `--version`

File: tests/test_shtab.py [hunk 3]
```diff
@@ -108,6 +108,8 @@ def test_prog_scripts(shell, caplog, capsys):
             "#compdef script.py",
             "_describe 'script.py commands' _commands",
         ]
+    elif shell == "tcsh":
+        assert script_py == ["complete script.py \\"]
```
Comment (casperdcl) [inline tests/test_shtab.py:112]: :sweat_smile: doesn't look like it's doing anything?

**Review by casperdcl — COMMENTED** [review]
thanks! Just not sure about the test (seems to imply `complete_tcsh` doesn't really do much?)

File: tests/test_shtab.py (hunk 3 above)
Comment (simaoafonso-pwt) [reply tests/test_shtab.py:112]: Ah, now I understand what you mean. As you can see on the top comment, I generate multiple lines separated by `\\`, the line continuation. That's why it seems it's not doing anything. :grin: 

I could put it all in a giant line, but that hurts readability, one of the (good) ideas behind the project.

File: examples/customcomplete.py [hunk 4]
```diff
@@ -57,6 +58,7 @@ def get_main_parser():
             " accidentally overwriting existing files."
         ),
     ).complete = shtab.DIRECTORY  # directory tab completion builtin shortcut
+    main_parser.add_argument('ochoice', choices=['oa', 'ob'], help='Other Choices')
```
[the remark this answers is not in the document]
Comment (simaoafonso-pwt) [reply examples/customcomplete.py:61]: I just added this to debug the sub-parsers. This is a second choice for the root parser, not to be confused with the first choice of the first sub parser.

It works well for the bash autocomplete too.

Just a little torture test, but it's an edge case, I can take it out.

[the remark this answers is not in the document]
Comment (casperdcl) [reply examples/customcomplete.py:61]: hmm maybe this instead...

```suggestion
    main_parser.add_argument(
        "--suffix", choices=["json", "csv"], default="json", help="Output format",
    )
```

File: tests/test_shtab.py (hunk 3 above)
Comment (casperdcl) [reply tests/test_shtab.py:112]: sure I guess the tests are fairly basic anyway :)

**Comment by casperdcl** [comment]
> I can squash everything into a single commit in the end, if you prefer.

no need, it's ok :)

**Review by casperdcl — COMMENTED** [review]
awesome! can you update `docs/use.md` (used to generate https://docs.iterative.ai/shtab/use/) with installation instructions for `tcsh`?

File: examples/customcomplete.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (simaoafonso-pwt) [reply examples/customcomplete.py]: What I was testing was in fact having a second positional argument on the root parser. When this is changed to an optional argument, it's not testing the same thing.

But this is an esoteric thing anyway, it's your call.

[the remark this answers is not in the document]
Comment (simaoafonso-pwt) [reply examples/customcomplete.py]: > But this is an esoteric thing anyway, it's your call.

Revert 4084ce29dabeaf7f3d7c65b034d75676f1ae38b1 if you want to use your version.

File: docs/use.md [hunk 5]
```diff
@@ -77,6 +77,30 @@ Below are various examples of enabling `shtab`'s own tab completion scripts.
     shtab --shell=zsh shtab.main.get_main_parser > ~/.zsh/completions/_shtab
     ```
 
+=== "tcsh"
+
+    `tcsh` does not have support for lazy loading shell completion, this will
+    need to be loaded just like any other initialization script.
+
+    ```csh
```
Comment (casperdcl) [inline docs/use.md:85]:
```suggestion
    ```sh
```

File: docs/use.md [hunk 6]
```diff
@@ -77,6 +77,30 @@ Below are various examples of enabling `shtab`'s own tab completion scripts.
...
+      | sudo tee /etc/profile.d/shtab.completion.csh
+    ```
+
+=== "Eager tcsh"
+
+    There are a few options:
+
+    ```sh
+    # Install locally
+    echo 'shtab --shell=tcsh shtab.main.get_main_parser | source /dev/stdin` \
```
Comment (casperdcl) [inline docs/use.md:96]: typo:

```suggestion
    echo 'shtab --shell=tcsh shtab.main.get_main_parser | source /dev/stdin' \
```

File: docs/use.md [hunk 7]
```diff
@@ -77,6 +77,30 @@ Below are various examples of enabling `shtab`'s own tab completion scripts.
     shtab --shell=zsh shtab.main.get_main_parser > ~/.zsh/completions/_shtab
     ```
 
+=== "tcsh"
+
+    `tcsh` does not have support for lazy loading shell completion, this will
+    need to be loaded just like any other initialization script.
```
Comment (casperdcl) [inline docs/use.md:83]:
```suggestion
```

**Review by casperdcl — CHANGES_REQUESTED** [review]
minor improvements

**Review by casperdcl — COMMENTED** [review]
awesome, thanks!

**Comment by simaoafonso-pwt** [comment]
Thanks for the quick review.

**Comment by simaoafonso-pwt** [comment]
https://github.com/iterative/shtab/runs/4227856457?check_suite_focus=true#step:5:31

Oops, I did not run the tests on python 2.7, and nether did the actions run when I pushed.

https://github.com/iterative/shtab/blob/46302c5ca80d6eeab95f328cf85e152e55a32d79/shtab/__init__.py#L644

https://github.com/iterative/shtab/blob/46302c5ca80d6eeab95f328cf85e152e55a32d79/shtab/__init__.py#L661

Unpacking lists doesn't work. Did a quick test, you can apply this patch:

```patch
diff --git i/shtab/__init__.py w/shtab/__init__.py
index d3747f7..da279b4 100644
--- i/shtab/__init__.py
+++ w/shtab/__init__.py
@@ -641,7 +641,7 @@ def complete_tcsh(parser, root_prefix=None, preamble="", choice_functions=None):
                 if not requirements and isinstance(positional.choices, dict):
                     for subcmd, subparser in positional.choices.items():
                         log.debug('%s| | SubParser: %s', log_prefix, subcmd)
-                        recurse_parser(subparser, positional_idx, [*requirements, subcmd])
+                        recurse_parser(subparser, positional_idx, [r for r in requirements] + [subcmd])

     recurse_parser(parser, 0)

@@ -658,7 +658,7 @@ def complete_tcsh(parser, root_prefix=None, preamble="", choice_functions=None):
                     '[ "$cmd[{}]" == "{}" ]'.format(iidx, n) for iidx, n in enumerate(nn, start=2)]
                 if arg.choices:
                     nlist.append('( {}echo "{}" || false )'.format(
-                        ' && '.join([*checks, '']),                 # Append the separator
+                        ' && '.join([c for c in checks] + ['']),                 # Append the separator
                         '\\n'.join(arg.choices),
                     ))
```

**Comment by casperdcl** [comment]
yes I think misconfigured tests also means they didn't run pre-merge

**Comment by simaoafonso-pwt** [comment]
> yes I think misconfigured tests also meant they didn't run pre-merge

The main problem is that it needs a new tag, since the existing tag failed halfway. The tests pass with that tag.

**Comment by simaoafonso-pwt** [comment]
Oops, I was testing this on a large parser, and I missed skipping the `argparse.SUPRESS` optional arguments. The positional are correctly skipped.

https://github.com/iterative/shtab/blob/3aca48786fc8b2d8d88f75902cdfebabbf7ef1c5/shtab/__init__.py#L628-L634

It's a matter of gating that with `if optional.help != SUPRESS:`. Should I open a PR for this oneline change?

**Comment by casperdcl** [comment]
> Should I open a PR for this oneline change?

yes please do :pray:
